libraries/AR_WPNav/AR_WPNav.cpp

```diff
@@ -0,0 +1,371 @@
+    }
+
+    // if error is larger than pivot_turn_angle then use pivot steering at next WP
+    if (fabsf(yaw_error_cd) * 0.01f > _pivot_angle) {
+        return true;
+    }
+
+    return false;
+}
+
+// returns true if vehicle should pivot immediately (because heading error is too large)
+bool AR_WPNav::use_pivot_steering(float yaw_error_cd)
+{
+    // check cases where we clearly cannot use pivot steering
+    if (!_pivot_possible || (_pivot_angle <= 0)) {
+        _pivot_active = false;
+        return false;
+    }
+
+    // calc bearing error
+    const float yaw_error = fabsf(yaw_error_cd) * 0.01f;
+
+    // if error is larger than pivot_turn_angle start pivot steering
+    if (yaw_error > _pivot_angle) {
+        _pivot_active = true;
+        return true;
+    }
+
+    // if within 10 degrees of the target heading, exit pivot steering
+    if (yaw_error < 10.0f) {
+        _pivot_active = false;
+        return false;
+    }
+
+    // by default stay in
+    return _pivot_active;
+}
+
+// true if update has been called recently
+bool AR_WPNav::is_active() const
+{
+    return ((AP_HAL::millis() - _last_update_ms) < AR_WPNAV_TIMEOUT_MS);
+}
+
+// update distance from vehicle's current position to destination
+void AR_WPNav::update_distance_and_bearing_to_destination()
+{
+    // if no current location leave distance unchanged
+    Location current_loc;
+    if (!_orig_and_dest_valid || !AP::ahrs().get_position(current_loc)) {
+        _distance_to_destination = 0.0f;
+        _wp_bearing_cd = 0.0f;
+        return;
+    }
+    _distance_to_destination = current_loc.get_distance(_destination);
+    _wp_bearing_cd = current_loc.get_bearing_to(_destination);
+}
+
+// calculate steering output to drive along line from origin to destination waypoint
+// relies on update_distance_and_bearing_to_destination being called first so _wp_bearing_cd has been updated
+void AR_WPNav::update_steering(const Location& current_loc, float current_speed)
+{
+    // calculate yaw error for determining if vehicle should pivot towards waypoint
+    float yaw_cd = _reversed ? wrap_360_cd(_wp_bearing_cd + 18000) : _wp_bearing_cd;
+    float yaw_error_cd = wrap_180_cd(yaw_cd - AP::ahrs().yaw_sensor);
+
+    // calculate desired turn rate and update desired heading
+    if (use_pivot_steering(yaw_error_cd)) {
+        _cross_track_error = 0.0f;
+        _desired_heading_cd = yaw_cd;
+        _desired_lat_accel = 0.0f;
+        _desired_turn_rate_rads = _atc.get_turn_rate_from_heading(radians(yaw_cd * 0.01f), radians(_pivot_rate));
+    } else {
+        // run L1 controller
+        _nav_controller.set_reverse(_reversed);
+        _nav_controller.update_waypoint((_reached_destination ? current_loc : _origin), _destination, _radius);
+
+        // retrieve lateral acceleration, heading back towards line and crosstrack error
+        float desired_lat_accel = constrain_float(_nav_controller.lateral_acceleration(), -_turn_max_mss, _turn_max_mss);
```

Comment:
The reason I did this is that I somehow think we shouldn't put incomplete values into an externally accessible variable.  I don't know of any cases where other threads would be accessing the value (except maybe mavlink?) but I felt like keeping it slightly more atomic was a good thing.  I suspect it doesn't actually help much though.

Reply:
Understood. I don't think even MAVLink is in another thread looking at this (otherwise we would need locks here), but what prompted me to notice this was the `_desired_heading_cd` line below is being updated directly, even though it might also not the final value.

Reply:
yes, you're right.. it's inconsistent.. I'll remove the temp variable, txs!